src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,315 @@
+	// To address this issue, we ensure that every incoming regular expression is modified to be anchored
+	// when querying the map-backed segment, and isn't anchored when querying Vellum's RE.
+	simpleRe, err := ensureRegexpAnchored(vellumRe)
+	if err != nil {
+		return CompiledRegex{}, fmt.Errorf("unable to create map re: %v", err)
+	}
+
+	var (
+		compiledRegex = CompiledRegex{}
+	)
+	simpleRE, err := re.Compile(simpleRe.String())
+	if err != nil {
+		return compiledRegex, err
+	}
+	compiledRegex.Simple = simpleRE
+
+	fstRE, start, end, err := fstregexp.ParsedRegexp(reString, vellumRe)
+	if err != nil {
+		return compiledRegex, err
+	}
+	compiledRegex.FST = fstRE
+	compiledRegex.PrefixBegin = start
+	compiledRegex.PrefixEnd = end
+
+	return compiledRegex, nil
+}
+
+func parseRegexp(re string) (*syntax.Regexp, error) {
+	return syntax.Parse(re, syntax.Perl)
+}
+
+// ensureRegexpAnchored adds '^' and '$' characters to appropriate locations in the parsed syntax.Regexp,
+// to ensure every input regular expression is converted to it's equivalent anchored regular expression.
+// NB: assumes input regexp AST is un-anchored.
+func ensureRegexpAnchored(unanchoredRegexp *syntax.Regexp) (*syntax.Regexp, error) {
+	ast := &syntax.Regexp{
+		Op:    syntax.OpConcat,
+		Flags: syntax.Perl,
+		Sub: []*syntax.Regexp{
+			&syntax.Regexp{
+				Op:    syntax.OpBeginText,
+				Flags: syntax.Perl,
+			},
+			unanchoredRegexp,
+			&syntax.Regexp{
+				Op:    syntax.OpEndText,
+				Flags: syntax.Perl,
+			},
+		},
+	}
+	return simplify(ast), nil
+}
+
+// ensureRegexpUnanchored strips '^' and '$' characters from appropriate locations in the parsed syntax.Regexp,
+// to ensure every input regular expression is converted to it's equivalent un-anchored regular expression
+// assuming the entire input is matched.
+func ensureRegexpUnanchored(parsed *syntax.Regexp) (*syntax.Regexp, error) {
+	r, _, err := ensureRegexpUnanchoredHelper(parsed, true, true)
+	if err != nil {
+		return nil, err
+	}
+	return simplify(r), nil
+}
+
+func ensureRegexpUnanchoredHelper(parsed *syntax.Regexp, leftmost, rightmost bool) (output *syntax.Regexp, changed bool, err error) {
+	changed = false
+	switch parsed.Op {
+	case syntax.OpBeginLine, syntax.OpEndLine:
+		// i.e. the flags provided to syntax.Parse did not include the `OneLine` flag, which
+		// should never happen as we're using syntax.Perl which does include it (ensured by a test
+		// in this package).
+		return nil, false, fmt.Errorf("regular expressions are forced to be single line")
+	case syntax.OpBeginText:
+		if leftmost {
+			return &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}, true, nil
+		}
+		return parsed, false, nil
+	case syntax.OpEndText:
+		if rightmost {
+			return &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}, true, nil
+		}
+		return parsed, false, nil
```

Comment:
Hm it's strange that some of these are early returns and others alter the vars and let the bottom of the switch case return.   Could we consolidate on one?

Reply:
sure thing

Reply:
done